src/plugins/kdb/lmdb/kdb_lmdb.c

```diff
@@ -0,0 +1,1142 @@
+        txn = temp_txn;
+    }
+
+    if (must_overwrite && mdb_get(txn, db, &key, &dummy) == MDB_NOTFOUND) {
+        mdb_txn_abort(temp_txn);
+        return KRB5_KDB_NOENTRY;
+    }
+
+    err = mdb_put(txn, db, &key, &val, putflags);
+    if (err)
+        goto error;
+
+    if (temp_txn != NULL) {
+        err = mdb_txn_commit(temp_txn);
+        temp_txn = NULL;
+        if (err)
+            goto error;
+    }
+
+    return 0;
+
+error:
+    mdb_txn_abort(temp_txn);
+    if (err == MDB_KEYEXIST)
+        return KRB5_KDB_INUSE;
+    else
+        return klerr(context, err, _("LMDB write failure"));
+}
+
+/* Delete an entry from the specified env and database, using a temporary write
+ * transaction.  Return KRB5_KDB_NOENTRY if the key does not exist. */
+static krb5_error_code
+del(krb5_context context, MDB_env *env, MDB_dbi db, char *keystr)
+{
+    krb5_error_code ret = 0;
+    MDB_txn *txn;
+    MDB_val key = { strlen(keystr), keystr };
+    int err;
+
+    err = mdb_txn_begin(env, NULL, 0, &txn);
+    if (!err)
+        err = mdb_del(txn, db, &key, NULL);
+    if (!err) {
+        err = mdb_txn_commit(txn);
+        txn = NULL;
+    }
+
+    if (err == MDB_NOTFOUND)
+        ret = KRB5_KDB_NOENTRY;
+    else if (err)
+        ret = klerr(context, err, _("LMDB delete failure"));
+
+    mdb_txn_abort(txn);
+    return ret;
+}
+
+/* Zero out and unlink filename. */
+static krb5_error_code
+destroy_file(const char *filename)
+{
+    krb5_error_code ret;
+    struct stat st;
+    ssize_t len;
+    off_t pos;
+    uint8_t buf[BUFSIZ], zbuf[BUFSIZ] = { 0 };
+    int fd;
+
+    fd = open(filename, O_RDWR, 0);
+    if (fd < 0)
+        return errno;
+    set_cloexec_fd(fd);
+    if (fstat(fd, &st) == -1)
+        goto error;
+
+    memset(zbuf, 0, BUFSIZ);
+    pos = 0;
+    while (pos < st.st_size) {
+        len = read(fd, buf, BUFSIZ);
+        if (len < 0)
+            goto error;
+        /* Only rewrite the block if it's not already zeroed, in case the file
+         * is sparse. */
+        if (memcmp(buf, zbuf, len) != 0) {
+            (void)lseek(fd, pos, SEEK_SET);
+            len = write(fd, zbuf, len);
+            if (len < 0)
+                goto error;
+        }
+        pos += len;
+    }
+    close(fd);
+
+    if (unlink(filename) != 0)
+        return errno;
+    return 0;
+
+error:
+    ret = errno;
+    close(fd);
+    return ret;
+}
+
+static krb5_error_code
+klmdb_lib_init()
+{
+    return 0;
+}
+
+static krb5_error_code
+klmdb_lib_cleanup()
+{
+    return 0;
+}
+
+static krb5_error_code
+klmdb_fini(krb5_context context)
+{
+    klmdb_context *dbc;
+
+    if (context->dal_handle->db_context == NULL)
+        return 0;
+    dbc = context->dal_handle->db_context;
+    mdb_txn_abort(dbc->read_txn);
+    mdb_txn_abort(dbc->load_txn);
+    mdb_env_close(dbc->env);
+    mdb_env_close(dbc->lockout_env);
+    free(dbc->path);
+    free(dbc->lockout_path);
+    free(dbc);
+    context->dal_handle->db_context = NULL;
```

Comment:
`dbc = NULL`

Reply:
That wouldn't have the same effect (or any effect, since it's right before a return statement).

Reply:
Right, got too excited.